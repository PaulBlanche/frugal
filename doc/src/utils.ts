type ThrottleOptions = {
  timeout: number;
  leading?: boolean;
  trailing?: boolean;
};

export function throttle<ARGS extends any[]>(
  callback: (...args: ARGS) => void,
  { timeout, leading = false, trailing = false }: ThrottleOptions,
): (...args: ARGS) => void {
  let timeoutHandle: number | undefined = undefined;
  let previous = 0;

  return (...args) => {
    const now = Date.now();

    if (!previous && !leading) {
      previous = now;
    }

    const remaining = timeout - (now - previous);

    if (remaining <= 0) {
      clearTimeout(timeoutHandle);
      timeoutHandle = undefined;
      previous = now;
      callback(...args);
    } else if (timeoutHandle === undefined && trailing) {
      timeoutHandle = setTimeout(() => {
        previous = Date.now();
        timeoutHandle = undefined;
        callback(...args);
      }, remaining);
    }
  };
}
