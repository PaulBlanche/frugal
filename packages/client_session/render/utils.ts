import { NodeType } from './types.ts';

export function clone(node: Node) {
    if (isScriptElement(node)) {
        // for scripts cloning is not enough, we need to recreate the script
        // element for it to be loaded/evaluated again
        const script = document.createElement('script');
        for (const attribute of node.attributes) {
            script.setAttribute(attribute.name, attribute.value);
        }
        script.innerHTML = node.innerHTML;
        return script;
    }
    return node.cloneNode(true);
}

function isScriptElement(node: Node): node is HTMLScriptElement {
    return isElement(node) && node.nodeName === 'SCRIPT';
}

export function isElement(node: Node): node is Element {
    return node.nodeType === NodeType.ELEMENT_NODE;
}

export function hash(
    element: Element,
    attributeFilter?: string[],
): string {
    const key = [];

    if (attributeFilter) {
        for (const name of attributeFilter) {
            const value = element.getAttribute(name);
            if (value) {
                key.push(`${name}="${element.getAttribute(name)}"`);
            }
        }
    } else {
        for (const attribute of element.attributes) {
            key.push(`${attribute.name}="${attribute.value}"`);
        }
    }

    key.sort();

    return `${element.tagName} ${key.join(' ')} ${element.innerHTML}`;
}

type Attribute = [name: string, value: string | boolean];

const BOOLEAN_ATTRIBUTES = [
    'allowfullscreen',
    'async',
    'autofocus',
    'autoplay',
    'checked',
    'controls',
    'default',
    'defer',
    'disabled',
    'formnovalidate',
    'inert',
    'ismap',
    'itemscope',
    'loop',
    'multiple',
    'muted',
    'nomodule',
    'novalidate',
    'open',
    'playsinline',
    'readonly',
    'required',
    'reversed',
    'selected',
];

export function getAttributes(element: Element): Attribute[] {
    const attributes: Record<string, string | boolean> = {};

    for (const { name, value } of element.attributes) {
        attributes[name] = value;
    }

    for (const name of BOOLEAN_ATTRIBUTES) {
        attributes[name] = element.hasAttribute(name);
    }

    return Object.entries(attributes);
}
