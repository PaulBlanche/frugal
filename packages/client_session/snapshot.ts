import * as murmur from '../murmur/mod.ts';

type Node = {
    name: string;
    attributes: { name: string; value: string }[];
    content: string;
    hash: string;
    element: Element;
};

export type Head = Node[];

export function serialize(head: HTMLHeadElement) {
    const snapshot: Head = [];

    head.childNodes.forEach((node) => {
        if (!(node instanceof Element)) {
            return;
        }

        const hash = new murmur.Hash();

        const name = node.nodeName.toLowerCase();
        const content = node.textContent ?? '';

        const sorted: string[] = [];
        const attributes: { name: string; value: string }[] = [];
        for (let i = 0, length = node.attributes.length; i < length; i++) {
            const attribute = node.attributes[i];
            attributes.push({
                name: attribute.name,
                value: attribute.value,
            });
            sorted.push(`${attribute.name}="${attribute.value}"`);
        }

        hash.update(name);

        sorted.sort();
        sorted.forEach((value) => {
            hash.update(value);
        });

        hash.update(content);

        snapshot.push({
            name,
            attributes,
            content,
            hash: hash.digest(),
            element: node,
        });
    });

    return snapshot;
}
