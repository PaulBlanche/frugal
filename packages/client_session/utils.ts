export function isInternalUrl(url: URL) {
    const rootURL = new URL('/', document.baseURI);
    return url.href.startsWith(rootURL.href);
}

export function getClosestParentNavigableAnchor(
    target: EventTarget,
): HTMLAnchorElement | undefined {
    if (target instanceof Element) {
        const anchor = target.closest<HTMLAnchorElement>(
            'a[href]:not([target^=_]:not([download])',
        );
        if (anchor !== null) {
            return anchor;
        }
    }
}

export function getAnchorUrl(anchor: HTMLAnchorElement) {
    return new URL(anchor.href, document.baseURI);
}

const readyStateOrder: Record<DocumentReadyState, number> = {
    'loading': 0,
    'interactive': 1,
    'complete': 2,
};

declare global {
    interface WindowEventMap {
        'frugal:readystatechange': CustomEvent<
            { readystate: DocumentReadyState }
        >;
    }
}

export function onReadyStateChange(
    readyState: DocumentReadyState,
    callback: () => void,
) {
    if (readyStateOrder[document.readyState] >= readyStateOrder[readyState]) {
        callback();
    } else {
        document.addEventListener('readystatechange', () => {
            if (document.readyState === readyState) {
                callback();
            }
        });
    }

    addEventListener(
        'frugal:readystatechange',
        (event) => {
            if (event.detail.readystate === readyState) {
                callback();
            }
        },
    );
}
