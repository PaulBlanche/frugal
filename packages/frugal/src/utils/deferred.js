import * as _type from "./_type/deferred.js";
export * from "./_type/deferred.js";

/**
 * @template T
 * @returns {_type.Deferred<T>}
 */
export function deferred() {
    let methods;
    let state = "pending";
    const promise = new Promise((resolve, reject) => {
        methods = {
            /** @param {T | PromiseLike<T>} value */
            async resolve(value) {
                await value;
                state = "fulfilled";
                resolve(value);
            },
            /** @param {any} reason */
            reject(reason) {
                state = "rejected";
                reject(reason);
            },
        };
    });
    Object.defineProperty(promise, "state", { get: () => state });
    return Object.assign(promise, methods);
}
