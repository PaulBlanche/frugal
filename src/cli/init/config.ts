import * as path from "../../../dep/std/path.ts";
import * as dotenv from "../../../dep/std/flags.ts";
import * as colors from "../../../dep/std/fmt/colors.ts";
import { CacheType, ExportType, Framework, toCacheType, toExportType, toFramework, VALID_CACHE } from "./types.ts";

export type Config = {
    root: string;
    exportType: ExportType;
    cacheType: CacheType;
    framework: Framework;
    force: boolean;
};

export function getConfig(args: string[]): Config {
    const flags = dotenv.parse(args, {
        string: ["export", "cache", "framework"],
        boolean: ["force", "help"],
        default: {
            "force": false,
        },
    });

    if (flags._[0]) {
        const exportType = toExportType(flags.export);
        if (exportType === undefined) {
            throw Error(
                `invalid export type "${flags.export}". Valid values are: ${Object.values(ExportType).join(", ")}`,
            );
        }

        const cacheType = toCacheType(flags.cache);
        if (cacheType === undefined || VALID_CACHE[exportType].includes(cacheType)) {
            throw Error(
                `invalid cache type "${flags.cache}". Valid values are: ${VALID_CACHE[exportType].join(", ")}`,
            );
        }

        const framework = toFramework(flags.framework);
        if (framework === undefined) {
            throw Error(
                `invalid cache type "${flags.framework}". Valid values are: ${Object.values(Framework).join(", ")}`,
            );
        }

        const root = path.resolve(String(flags._[0]));

        if (!isEmpty(root) && !flags.force) {
            console.error(`The directory ${root} is not empty. Use --force to overwrite it.`);
            Deno.exit(1);
        }

        return {
            root,
            exportType,
            cacheType,
            framework,
            force: flags.force,
        };
    }

    const name = askUntil("Enter a project name", {
        default: "frugal-project",
    });
    const root = path.resolve(name);

    const force = isEmpty(root) ||
        confirm(colors.yellow(`The directory ${root} is not empty. Do you want to overwrite it ?`));

    if (!isEmpty(root) && !force) {
        console.error(`The directory is not empty.`);
        Deno.exit(1);
    }

    const exportType = askUntil("Choose an export type", {
        head: "Projects have to be exported for specific plateform to be deployed.",
        default: ExportType.NO_EXPORT,
        values: Object.values(ExportType),
        sanitize: (exportType) => toExportType(exportType),
    });

    const cacheType = VALID_CACHE[exportType].length === 1
        ? VALID_CACHE[exportType][0]
        : askUntil("Choose an cache type", {
            head: "Frugal will have to cache static pages in some storage.",
            default: VALID_CACHE[exportType][0],
            values: VALID_CACHE[exportType],
            sanitize: (cacheType) => toCacheType(cacheType),
        });

    const framework = askUntil("Choose an framework", {
        head: "Frugal works with different frameworks.",
        default: Framework.NO_FRAMEWORK,
        values: Object.values(Framework),
        sanitize: (exportType) => toFramework(exportType),
    });

    return {
        root,
        exportType,
        cacheType,
        framework,
        force,
    };
}

function ask(message: string): string | undefined;
function ask(message: string, _default: string): string;
function ask(message: string, _default?: string): string | undefined;
function ask(message: string, _default?: string): string | undefined {
    const value = prompt(message, _default ? colors.gray(_default) : undefined);
    if (value === null) {
        return undefined;
    }
    return colors.stripColor(value);
}

type AskUntilConfig<VALUE> = {
    head?: string;
    default?: string;
    values?: string[];
    sanitize?: (userValue: string | undefined) => VALUE | undefined;
};

function askUntil<VALUE = string>(
    message: string,
    { head, default: _default, values, sanitize = (x) => (x as any) }: AskUntilConfig<VALUE>,
): VALUE {
    const askMessageParts = [colors.bold(message)];
    if (values) {
        askMessageParts.push(`(${values.join(", ")})`);
    }

    const askMessage = askMessageParts.join(" ");

    console.log("");
    head && console.log(colors.gray(`>> ${head}`));

    while (true) {
        const userValue = ask(askMessage, _default);

        const sanitizedValue = sanitize(userValue);

        if (sanitizedValue !== undefined) {
            return sanitizedValue;
        }

        const errorMessageParts = [`invalid value "${userValue}".`];
        if (values) {
            errorMessageParts.push(`Valid values are: ${values.join(", ")}`);
        }

        console.log(colors.red(errorMessageParts.join(" ")));
    }
}

function isEmpty(path: string) {
    try {
        const dir = [...Deno.readDirSync(path)];
        const isEmpty = dir.length === 0 ||
            dir.length === 1 && dir[0].name === ".git";

        return isEmpty;
    } catch (error) {
        if ((error instanceof Deno.errors.NotFound)) {
            return true;
        }
        throw error;
    }
}
