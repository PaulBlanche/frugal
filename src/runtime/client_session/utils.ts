export function isInternalUrl(url: URL) {
    const rootURL = new URL('/', document.baseURI);
    return url.href.startsWith(rootURL.href);
}

export function getClosestParentNavigableAnchor(
    target: EventTarget,
): HTMLAnchorElement | undefined {
    if (target instanceof Element) {
        const anchor = target.closest<HTMLAnchorElement>(
            'a[href]:not([target^=_]:not([download])',
        );
        if (anchor !== null) {
            return anchor;
        }
    }
}

export function getUrl(path: string) {
    return new URL(path, document.baseURI);
}

const readyStateOrder: Record<DocumentReadyState, number> = {
    'loading': 0,
    'interactive': 1,
    'complete': 2,
};

declare global {
    type FrugalReadyStateChangeEvent = CustomEvent<
        { readystate: DocumentReadyState }
    >;

    interface WindowEventMap {
        'frugal:readystatechange': FrugalReadyStateChangeEvent;
    }
}

export function onReadyStateChange(
    readyState: DocumentReadyState,
    callback: () => void,
) {
    if (typeof document === 'undefined') {
        return;
    }

    if (readyStateOrder[document.readyState] >= readyStateOrder[readyState]) {
        callback();
    } else {
        document.addEventListener('readystatechange', () => {
            if (document.readyState === readyState) {
                callback();
            }
        });
    }

    addEventListener(
        'frugal:readystatechange',
        (event) => {
            if (event.detail.readystate === readyState) {
                callback();
            }
        },
    );
}
