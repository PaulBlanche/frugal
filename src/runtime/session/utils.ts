export function isInternalUrl(url: URL) {
    const rootURL = new URL("/", document.baseURI);
    return url.href.startsWith(rootURL.href);
}

export function getClosestParentNavigableAnchor(
    target: EventTarget,
): HTMLAnchorElement | undefined {
    if (target instanceof Element) {
        const anchor = target.closest<HTMLAnchorElement>(
            "a[href]:not([target^=_]:not([download])",
        );
        if (anchor !== null) {
            return anchor;
        }
    }
}

export function getUrl(path: string) {
    return new URL(path, document.baseURI);
}
