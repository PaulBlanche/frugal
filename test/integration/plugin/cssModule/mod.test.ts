import * as asserts from "../../../../dep/std/testing/asserts.ts";
import * as snapshot from "../../../../dep/std/testing/snapshot.ts";
import { context } from "../../../../mod.ts";
import { BuildHelper } from "../../../utils.ts";

import { config } from "./frugal.config.ts";

if (import.meta.main) {
    await context(config).watch();
}

Deno.test("css: build page with css module dependencies", async (t) => {
    const helper = new BuildHelper(config);

    await helper.build();

    await new Promise((res) => setTimeout(res, 200));

    snapshot.assertSnapshot(t, await Deno.readTextFile(new URL("css/page.css", helper.config.publicdir)));

    const cache = await helper.cacheExplorer();

    snapshot.assertSnapshot(t, await cache.loadDocument("/page"));
});

Deno.test("css: css module dependencies are watched", async () => {
    const helper = new BuildHelper(config);

    const context = helper.context();
    context.watch();

    await context.awaitNextBuild();

    const firstBuildCache = await context.cacheExplorer();
    const updatedAt = firstBuildCache.get("/page").updatedAt;

    // add a comment at the top of dep.css
    const depModuleURL = new URL("./dep.module.css", import.meta.url);
    const originalData = await Deno.readTextFile(depModuleURL);
    await Deno.writeTextFile(depModuleURL, `\n${originalData}`);

    await context.awaitNextBuild();

    const secondBuildChache = await context.cacheExplorer();
    asserts.assertNotEquals(updatedAt, secondBuildChache.get("/page").updatedAt);

    await Deno.writeTextFile(depModuleURL, originalData);

    await context.dispose();
});
