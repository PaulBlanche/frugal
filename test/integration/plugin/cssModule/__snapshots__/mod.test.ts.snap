export const snapshot = {};

snapshot[`css: build page with css module dependencies 1`] = `
"/* virtual:/home/whiteshoulders/perso/frugal/test/integration/plugin/cssModule/main.module.css-92ca48743c13d85a.frugal-compiled-module.css */
" +
  "._5TXYoG_foo {
" +
  "  color: red;
" +
  "}
" +
  "._5TXYoG_bar {
" +
  "  background: red;
" +
  "}
" +
  "
" +
  "/* virtual:/home/whiteshoulders/perso/frugal/test/integration/plugin/cssModule/dep.module.css-f6b3d0584c4c3682.frugal-compiled-module.css */
" +
  ".K3qUyq_bar {
" +
  "  margin: 10px;
" +
  "}
"
`;

snapshot[`css: build page with css module dependencies 2`] = `'{"bar":["_5TXYoG_bar","K3qUyq_bar"],"foo":["_5TXYoG_foo","_5TXYoG_bar","K3qUyq_bar"]}'`;
